contrib/fixup-libgfortran.sh

```diff
@@ -40,65 +40,87 @@ find_shlib ()
+        LIBQUADMATH_SONAMES="$LIBQUADMATH_SONAMES $(basename "$LIBQUADMATH_PATH")"
     fi
 done
 
-LIBGFORTRAN_DIRS=$(echo "$LIBGFORTRAN_DIRS" | tr " " "\n" | sort | uniq | grep -v '^$' | tr "\n" " ")
+# Take in a list of space-separated tokens, return a deduplicated list of the same
+uniquify()
+{
+    echo "$1" | tr " " "\n" | sort | uniq | grep -v '^$' | tr "\n" " "
+}
 
-# If only we could agree on something
-if [ "$UNAME" = "Linux" ]; then
-    NAMEXTS="gcc_s.so.1 gfortran.so.3 quadmath.so.0"
-elif [ "$UNAME" = "Darwin" ]; then
-    NAMEXTS="gcc_s.1.dylib gfortran.3.dylib quadmath.0.dylib"
-fi
+LIBGFORTRAN_DIRS=$(uniquify "$LIBGFORTRAN_DIRS")
+SONAMES="$(uniquify "$LIBGFORTRAN_SONAMES $LIBGCC_SONAMES $LIBQUADMATH_SONAMES")"
 
-for namext in $NAMEXTS; do
+# Copy the SONAMEs we identified above into our private_libdir
+for soname in $SONAMES; do
     for dir in $LIBGFORTRAN_DIRS; do
-        if [ ! -f "$private_libdir/lib$namext" ] && [ -f "$dir/lib$namext" ]; then
-            cp -v "$dir/lib$namext" "$private_libdir"
-            chmod 755 "$private_libdir/lib$namext"
+        if [ ! -f "$private_libdir/$soname" ] && [ -f "$dir/$soname" ]; then
+            cp -v "$dir/$soname" "$private_libdir"
+            chmod 755 "$private_libdir/$soname"
             if [ "$UNAME" = "Darwin" ]; then
-                install_name_tool -id @rpath/lib$namext "$private_libdir/lib$namext"
+                install_name_tool -id "@rpath/$soname" "$private_libdir/$soname"
             fi
         fi
     done
-done
 
-# Add possible internal directories to LIBGFORTRAN_DIRS
-for lib in gfortran.3 quadmath.0 gcc_s.1 ; do
-    if [ -f "$private_libdir/lib$lib.$SHLIB_EXT" ]; then
+    # Add possible internal directories to LIBGFORTRAN_DIRS so that we can find all possible ways
+    # That any of our libraries or these libraries we just copied in link to themselves later.
+    if [ -f "$private_libdir/$soname" ]; then
         LIBGFORTRAN_DIRS="$LIBGFORTRAN_DIRS $(dirname $(find_shlib $lib libgfortran) 2>/dev/null)"
         LIBGFORTRAN_DIRS="$LIBGFORTRAN_DIRS $(dirname $(find_shlib $lib libgcc_s) 2>/dev/null)"
         LIBGFORTRAN_DIRS="$LIBGFORTRAN_DIRS $(dirname $(find_shlib $lib libquadmath) 2>/dev/null)"
     fi
 done
 
-LIBGFORTRAN_DIRS=$(echo "$LIBGFORTRAN_DIRS" | tr " " "\n" | sort | uniq | grep -v '^$' | tr "\n" " ")
+
+change_linkage()
+{
+    shlib_path="$1"
+    old_link="$2"
+    new_link="$3"
+    if [ "$UNAME" = "Darwin" ]; then
+        install_name_tool -change "$old_link" "$new_link" "$shlib_path"
+    else # $UNAME is "Linux", we only have two options, see above
+        patchelf --set-rpath \$ORIGIN "$shlib_path"
+    fi
+}
+
+LIBGFORTRAN_DIRS=$(uniquify "$LIBGFORTRAN_DIRS")
 echo "Found traces of libgfortran/libgcc in $LIBGFORTRAN_DIRS"
 
+# For every library that remotely touches libgfortran stuff (so the libraries
+# we have copied in ourselves as well as arpack, openlibm, etc...) we must
```

Comment:
openlibm shouldn't have any fortran in it these days, that was all moved to openspecfun